runtime/src/legacy_memory_manager/cpp/Memory.cpp

```diff
@@ -430,8 +437,70 @@ inline bool isShareable(ContainerHeader* container) {
     return container == nullptr || container->shareable();
 }
 
+MetaObjHeader* createMetaObject(TypeInfo** location);
+
+void destroyMetaObject(TypeInfo** location);
+
+bool has_meta_object(const ObjHeader* obj) {
+  auto* typeInfoOrMeta = clearPointerBits(obj->typeInfoOrMeta_, OBJECT_TAG_MASK);
+  return (typeInfoOrMeta != typeInfoOrMeta->typeInfo_);
+}
+
+MetaObjHeader* meta_object_for(ObjHeader* obj) {
+   return has_meta_object(obj) ?
+      reinterpret_cast<MetaObjHeader*>(clearPointerBits(obj->typeInfoOrMeta_, OBJECT_TAG_MASK)) :
+      createMetaObject(&obj->typeInfoOrMeta_);
+}
+
+void setContainerFor(ObjHeader* obj, ContainerHeader* container) {
+  meta_object_for(obj)->container_ = container;
+  obj->typeInfoOrMeta_ = setPointerBits(obj->typeInfoOrMeta_, OBJECT_TAG_NONTRIVIAL_CONTAINER);
+}
+
 }  // namespace
 
+ContainerHeader* containerFor(const ObjHeader* obj) {
+  unsigned bits = getPointerBits(obj->typeInfoOrMeta_, OBJECT_TAG_MASK);
+  if ((bits & (OBJECT_TAG_PERMANENT_CONTAINER | OBJECT_TAG_NONTRIVIAL_CONTAINER)) == 0)
+    return reinterpret_cast<ContainerHeader*>(const_cast<ObjHeader*>(obj)) - 1;
+  if ((bits & OBJECT_TAG_PERMANENT_CONTAINER) != 0)
+    return nullptr;
+  return (reinterpret_cast<MetaObjHeader*>(clearPointerBits(obj->typeInfoOrMeta_, OBJECT_TAG_MASK)))->container_;
+}
+
+ALWAYS_INLINE bool isPermanentOrFrozen(const ObjHeader* obj) {
+    auto* container = containerFor(obj);
+    return container == nullptr || container->frozen();
+}
+
+ALWAYS_INLINE bool isShareable(const ObjHeader* obj) {
+    auto* container = containerFor(obj);
+    return container == nullptr || container->shareable();
```

Comment:
AFAIK, none of the usages needs this `nullptr` check.

Reply:
Well, almost none.

Reply:
Removed here too.